sandbox/cgroup/kill.go

```diff
@@ -26,89 +26,184 @@ import (
+//     not created (e.g. on Ubuntu 18.04) so we use the freezer cgroup for tracking. This is
+//     only useful for killing apps or processes which do not have their lifecycle managed by
+//     external entities like systemd.
+//   - cgroup v2: stop forking through pids.max, kill processes until cgroup is drained.
+//     This is to address kernel versions without v2 freezer support so we use pids.max
+//     to prevent fork bombs from racing with snapd.
 var KillSnapProcesses = func(ctx context.Context, snapName string) error {
 	return errors.New("KillSnapProcesses not implemented")
 }
 
 var syscallKill = syscall.Kill
 
-// killProcessesInCgroup sends signal to all the processes belonging to
+var maxKillTimeout = 5 * time.Minute
+var killThawCooldown = 100 * time.Millisecond
+
+// killProcessesInCgroup sends SIGKILL signal to all the processes belonging to
 // passed cgroup directory.
 //
 // The caller is responsible for making sure that pids are not reused
 // after reading `cgroup.procs` to avoid TOCTOU.
-func killProcessesInCgroup(dir string, signal syscall.Signal) error {
-	pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
-	if err != nil {
-		return err
-	}
+//
+// The freeze() callback is called exactly before killing pids is started while the thaw()
+// callback is called exactly after killing pids ends and before returning errors to give
+// a chance to recover a cgroup from a frozen state. Not propagating errors from the
+// callbacks is intentional to make it clear that they are best effort.
+var killProcessesInCgroup = func(ctx context.Context, dir string, freeze func(ctx context.Context), thaw func()) error {
+	// Keep sending SIGKILL signals until no more pids are left in cgroup
+	// to cover the case where a process forks before we kill it.
+	for {
+		pids, err := pidsInFile(filepath.Join(dir, "cgroup.procs"))
+		if err != nil {
+			return err
+		}
+		if len(pids) == 0 {
+			// no more pids
+			return nil
+		}
 
-	var firstErr error
-	for _, pid := range pids {
-		pidNotFoundErr := syscall.ESRCH
-		if err := syscallKill(pid, signal); err != nil && !errors.As(err, &pidNotFoundErr) && firstErr == nil {
-			firstErr = err
+		if freeze != nil {
+			freeze(ctx)
+		}
+		var firstErr error
+		for _, pid := range pids {
+			// This prevents a rouge fork bomb from keeping this loop running forever
+			select {
+			case <-ctx.Done():
+				return fmt.Errorf("cannot kill processes in cgroup %q: %w", dir, ctx.Err())
+			default:
+			}
+
+			pidNotFoundErr := syscall.ESRCH
+			// TODO: Use pidfs when possible to avoid killing reused pids.
+			if err := syscallKill(pid, syscall.SIGKILL); err != nil && !errors.Is(err, pidNotFoundErr) && firstErr == nil {
+				firstErr = err
+			}
+		}
+		if thaw != nil {
+			// thaw() must be called before returning to avoid keeping cgroup stuck after freeze()
+			thaw()
+		}
+		if firstErr != nil {
+			return firstErr
```

Comment:
Have we ever encountered this in practice?

Reply:
No, according the [`man 2 kill`](https://linux.die.net/man/2/kill) it is unlikely because snapd is root